Docs/motion translate (#72)
bit-bots/bitbots_tools: bitbots_docs/docs/manual/tutorials/lowlevel.rst

Bitbots Lowlevel
================

What do the low level packages do?
----------------------------------

The low level packages are responsible for the motions of the robot.
For that, different hardware and software components are used; they are described in this article.
Additionally, common problems and troubleshooting strategies are explained.

The behavior of the low level packages is dominated by the control loop, i.e. a cycle that alternately reads and writes motor positions.
In parallel to this cycle, the new positions are calculated based on the observed values.

To react to problems as fast as possible, e.g. to overload errors in the motors, a faster control loop is desirable.
Formerly, our loop ran with a frequency of 100Hz, therefore the reaction time to errors it at least 20ms because the error has to be read, processed and written, which takes at least two cycles.

> There are essentially three possibilities to accelerate the control loop:

Author: all of them ;)

> There are essentially three possibilities to accelerate the control loop:

Reviewer: ```suggestion
There are essentially three possibilities which we use to accelerate the control loop:
```

Method 1. Type following (you should do this everytime when the usb once was plugged out or the connection was dropped)

.. code:: bash

    $ echo 1 | sudo tee /sys/bus/usb-serial/devices/ttyUSB0/latency_timer
    $ cat /sys/bus/usb-serial/devices/ttyUSB0/latency_timer    

Method 2. If you want to set it as be done automatically, and don't want to do above everytime, make rules file in /etc/udev/rules.d/. For example,

.. code:: bash

    $ echo ACTION==\"add\", SUBSYSTEM==\"usb-serial\", DRIVER==\"ftdi_sio\", ATTR{latency_timer}=\"1\" > 99-dynamixelsdk-usb.rules
    $ sudo cp ./99-dynamixelsdk-usb.rules /etc/udev/rules.d/
    $ sudo udevadm control --reload-rules
    $ sudo udevadm trigger --action=add
    $ cat /sys/bus/usb-serial/devices/ttyUSB0/latency_timer

If you have another good idea that can be an alternatives, Robotis is asking for advice via a github issue:  https://github.com/ROBOTIS-GIT/DynamixelSDK/issues

Dynamixel SDK
~~~~~~~~~~~~~

The Dynamixel SDK implements the Dynamixel protocol.
It provides methods to send instructions and to read status packets in different programming languages.

> We use a fork of Robotis' Dynamixel SDK because Robotis did not implement the sync read on multiple registers.

Reviewer: link the fork?

Dynamixel Workbench
~~~~~~~~~~~~~~~~~~~

The Dynamixel Workbench provides higher level functions than the Dynamixel SDK.
For example, the motor positions in the SDK are given as values between 0 and 4096 (2 Byte) which is converted to radians by the Dynamixel Workbench.
Thereby, the Workbench eases the work with the motors on a more abstract level.

ROS Control Framework
~~~~~~~~~~~~~~~~~~~~~

The ROS Control Framework is a part of ROS that is responsible for the motor and sensor control.
There are controllers for ROS Control that provide the interface between ROS and low level software parts.
These controllers are hardware agnostic because they are using interfaces to abstract from the hardware (e.g. motors).
To control the motors, the Dynamixel Controller is used, which itself uses the Dynamixel Hardware Interface.

ROS messages
~~~~~~~~~~~~

After all these steps we're finally at the ROS message level.

> There are two message types handled here: The joint state message gives the current positions of the motors, while the joint goals message can specify target positions for motors. The hardware interface also manages the IMU data and the values returned by the foot sensors.

Reviewer: These sentences should probably be each on their own line.

How do we use bitbots_ros_control?
----------------------------------

The package bitbots_ros_control provides the hardware interface for the dynamixel motors.

> The most important configuration file for this is the wolfgang.yaml. In this file you find multiple settings for defining, which values should be read from the motors (temperature, speed, force, ...), which sensors should be used (foot pressure sensors, IMU) and which settings should be set (control loop frequency, baud rate, CORE ports, auto torque, ...).

Reviewer: These sentences should probably be each on their own line.

The corresponding ROS node can be launched with `roslaunch bitbots_ros_control ros_control.launch`. This will execute the following steps:

> 1. The motors are pinged in alphabetical order. This happens due to the way yaml files are read. This means the HeadPan motor (id 19) is read first, while the RShoulderRoll motor (id 3) is read last.

Reviewer: ```suggestion
1. The motors are pinged one after another in alphabetical order. This happens due to the way yaml files are read. This means the HeadPan motor (id 19) is read first, while the RShoulderRoll motor (id 3) is read last.
```

Reviewer: The sentences should probably each be on their own line.

2. The values from the config file are written into the RAM and ROM of the motors. These are values like speed or return delay time.

> 3. The message "Hardware interface init finished" appears.

Reviewer: ```suggestion
3. The message "Hardware interface init finished" is printed to the terminal.
```
i assume?

4. The control loop starts, alternating between sync read and sync write.
5. The controllers for ROS control are loaded.

Help, I have a problem!
-----------------------

Error Opening Serial Port
~~~~~~~~~~~~~~~~~~~~~~~~~

> If you encounter the message "Error opening serial port", no connection between the NUC and the CORE board could be established. Therefore your first instinct should be checking whether the cable is plugged in correctly. If this does not solve the problem, you can check whether the board can be found by using `lsusb` (look at the "leaf" entry). You can further investigate this by using `ls /dev/`. You should find the devices "/dev/ttyUSB0" through "/dev/ttyUSB3", one for each of the four busses. If the names are different, you may have to alter the wolfgang.yaml file or unplug the CORE board and plug it back in, in order to make it use the known names.

Reviewer: These sentences should probably each be their own line.

Reviewer: ```suggestion
If you encounter the message "Error opening serial port", no connection between the NUC and the CORE board could be established. Therefore your first instinct should be checking whether the cable is plugged in correctly. If this does not solve the problem, you can check whether the board can be found by using `lsusb` (look for the "leaf" entry). You can further investigate this by using `ls /dev/`. You should find the devices "/dev/ttyUSB0" through "/dev/ttyUSB3", one for each of the four busses. If the names are different, you may have to alter the wolfgang.yaml file or unplug the CORE board and plug it back in, in order to make it use the known names.
```

> Motor problems

Reviewer: In this chapter the sentences should each have their own line

~~~~~~~~~~~~~~

> The first thing you should do if you have a motor problem ("no status from id", motors stuttering, ...) is checking whether all cables are plugged in correctly. Sometimes one of the cable sits loosely in its socket and may fall out entirely when the robot moves. To control whether all motors are reachable, this `software

Reviewer: ```suggestion
The first thing you should do if you have a motor problem ("no status from id", motors stuttering, ...) is checking whether all cables in the bus with the problem are plugged in correctly. Sometimes one of the cable sits loosely in its socket and may fall out entirely when the robot moves. To control whether all motors are reachable, this `software
```

Author: This edit should not be added, as even unrelated cables can cause issues in other motors. Of course it makes sense to check the affected cables first, but in general it makes sense to go through all cables, at least of that bus, to make sure.

Reviewer: ```suggestion
The first thing you should do if you have a motor problem ("no status from id", motors stuttering, ...) is checking whether all cables are plugged in correctly. Preferably starting with the bus with the motor that has a problem. Sometimes one of the cable sits loosely in its socket and may fall out entirely when the robot moves. To control whether all motors are reachable, this `software
```
